swim.go

```diff
@@ -63,28 +79,82 @@ func (s *SWIM) ShutDown() {
 //    The member(j) is judged to be failed, so check the member(j) as suspected or delete the member(j) from memberMap.
 //
 // ** When performing ping, ack, and indirect-ping in the above procedure, piggybackdata is sent together. **
+//
+//
+// startFailureDetector function
+//
+// 1. Pick a member from memberMap.
+// 2. Probe the member.
+// 3. After finishing probing all members, reset memberMap
+//
 func (s *SWIM) startFailureDetector() {
 
-}
+	go func() {
+		for {
+			// Get copy of current members from memberMap.
+			members := s.memberMap.GetMembers()
+			for _, member := range members {
+				s.probe(member)
+			}
 
-// Delegate interface to notify status change of the member
-type MemberStatusChangeDelegate interface {
-	deleteMember(id MemberID)
-	addMember()
-	updateMember()
+			// Reset memberMap.
+			s.memberMap.reset()
+		}
+	}()
+
+	<-s.quitFD
 }
 
-// Delete member
-func (s *SWIM) deleteMember(id MemberID) {
+// probe function
+//
+// 1. Send ping to the member(j) during the ack-timeout (time less than T).
+//    Return if ack message arrives on ack-timeout.
+//
+// 2. selects k number of members from the memberMap and sends indirect-ping(request k members to ping the member(j)).
+//    The nodes (that receive the indirect-ping) ping to the member(j) and ack when they receive ack from the member(j).
+//
+// 3. At the end of T, SWIM checks to see if ack was received from k members, and if there is no message,
+//    The member(j) is judged to be failed, so check the member(j) as suspected or delete the member(j) from memberMap.
+//
 
-}
+func (s *SWIM) probe(member Member) {
+
+	if member.Status == Dead {
+		return
+	}
+
+	end := make(chan bool, 1)
+	defer close(end)
+
+	go func() {
 
-// Add new member
-func (s *SWIM) addMember() {
+		// Ping to member
 
+		end <- true
+	}()
+
+	T := time.NewTimer(time.Duration(s.config.T))
+
+	select {
+	case <-end:
+		// Ended
+		return
+	case <-T.C:
+		// Suspect the member.
+		return
+	}
+}
+
+// Delegate interface to notify status change of the member
+type MemberMapUpdatedDelegate interface {
+	updateMember()
 }
 
 // Update member
+// 1. Check
+// 2.
+// 3.
+
 func (s *SWIM) updateMember() {
 
 }
```

Comment:
What is the intent of this function?

Reply:
I updated the comment!!

Reply:
Did you implement this function to separate the layer from `member_map.go` ?